File: demo/go/trade_demo.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"time"
)

const (
	urlStr    = "https://open-api.bingx.com"
	apiKey    = "Set your api key here!!"
	secretKey = "Set your secret key here!!"
)

func computeHmac256(strMessage string, strSecret string) string {
	key := []byte(strSecret)
	h := hmac.New(sha256.New, key)
	h.Write([]byte(strMessage))

	return hex.EncodeToString(h.Sum(nil))
}

func post(requestUrl string) (responseData []byte, err error) {
	u, _ := url.Parse(requestUrl)
	req := http.Request{
		Method: http.MethodPost,
		URL:    u,
		Header: http.Header{},
	}
	req.Header.Set("X-BX-APIKEY", apiKey)
	cli := http.Client{
		Timeout: 3 * time.Second,
	}
	response, err := cli.Do(&req)
	if err != nil {

		return
	}
	defer response.Body.Close()

	responseData, err = ioutil.ReadAll(response.Body)
	return
}

func placeOrder(symbol, side, tradeType, quantity, quoteOrderQty, price string) {
	requestPath := "/openApi/spot/v1/trade/order"
	requestUrl := urlStr + requestPath
	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
	mapParams := url.Values{}
	mapParams.Set("symbol", symbol)
	mapParams.Set("side", side)
	mapParams.Set("type", tradeType)
	mapParams.Set("quantity", quantity)
	mapParams.Set("quoteOrderQty", quoteOrderQty)
	mapParams.Set("price", price)
	mapParams.Set("timestamp", fmt.Sprint(timestamp))
	strParams := mapParams.Encode()
	strParams += "&signature=" + computeHmac256(strParams, secretKey)
	requestUrl += "?"
	requestUrl += strParams
	responseData, err := post(requestUrl)
	fmt.Println("\trequest:", requestUrl)
	fmt.Println("\tresponse:", string(responseData), err)
}

func cancelOrder(symbol, orderID string) {
	requestPath := "/openApi/spot/v1/trade/cancel"
	requestUrl := urlStr + requestPath
	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
	mapParams := url.Values{}
	mapParams.Set("symbol", symbol)
	mapParams.Set("orderId", orderID)
	mapParams.Set("timestamp", fmt.Sprint(timestamp))
	strParams := mapParams.Encode()
	strParams += "&signature=" + computeHmac256(strParams, secretKey)
	requestUrl += "?"
	requestUrl += strParams
	responseData, err := post(requestUrl)
	fmt.Println("\trequest:", requestUrl)
	fmt.Println("\tresponse:", string(responseData), err)
}

func queryOrder(symbol, orderID string) {
	requestPath := "/openApi/spot/v1/trade/query"
	requestUrl := urlStr + requestPath
	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
	mapParams := url.Values{}
	mapParams.Set("symbol", symbol)
	mapParams.Set("orderId", orderID)
	mapParams.Set("timestamp", fmt.Sprint(timestamp))
	strParams := mapParams.Encode()
	strParams += "&signature=" + computeHmac256(strParams, secretKey)
	requestUrl += "?"
	requestUrl += strParams
	responseData, err := post(requestUrl)
	fmt.Println("\trequest:", requestUrl)
	fmt.Println("\tresponse:", string(responseData), err)
}

func queryOpenOrders(symbol string) {
	requestPath := "/openApi/spot/v1/trade/openOrders"
	requestUrl := urlStr + requestPath
	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
	mapParams := url.Values{}
	mapParams.Set("symbol", symbol)
	mapParams.Set("timestamp", fmt.Sprint(timestamp))
	strParams := mapParams.Encode()
	strParams += "&signature=" + computeHmac256(strParams, secretKey)
	requestUrl += "?"
	requestUrl += strParams
	responseData, err := post(requestUrl)
	fmt.Println("\trequest:", requestUrl)
	fmt.Println("\tresponse:", string(responseData), err)
}

func queryHistoryOrders(symbol, orderId, startTime, endTime, limit string) {
	requestPath := "/openApi/spot/v1/trade/historyOrders"
	requestUrl := urlStr + requestPath
	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
	mapParams := url.Values{}
	mapParams.Set("symbol", symbol)
	mapParams.Set("orderId", orderId)
	mapParams.Set("startTime", startTime)
	mapParams.Set("endTime", endTime)
	mapParams.Set("limit", limit)
	mapParams.Set("timestamp", fmt.Sprint(timestamp))
	strParams := mapParams.Encode()
	strParams += "&signature=" + computeHmac256(strParams, secretKey)
	requestUrl += "?"
	requestUrl += strParams
	responseData, err := post(requestUrl)
	fmt.Println("\trequest:", requestUrl)
	fmt.Println("\tresponse:", string(responseData), err)
}

func main() {
	//fmt.Println("placeOrder:")
	//placeOrder("XRP-USDT", "BUY", "LIMIT", "20", "0", "0.5")

	//fmt.Println("cancelOrder:")
	//cancelOrder("XRP-USDT", "1513803323449278464")

	//fmt.Println("queryOrder:")
	//queryOrder("XRP-USDT", "1513804323790782464")

	//fmt.Println("queryOpenOrders:")
	//queryOpenOrders("XRP-USDT")

	//fmt.Println("queryHistoryOrders:")
	//queryHistoryOrders("XRP-USDT", "0", "0", "0", "2")
}

File: demo/php/trade_demo.php
<?php

$url = "https://open-api.bingx.com";
$apiKey = "Set your api key here!!";
$secretKey = "Set your secret key here!!";

function getOriginString(array $params) {
    // combine origin string
    $originString = "";
    $first = true;
    foreach($params as $n => $v) {
          if (!$first) {
              $originString .= "&";
          }
          $first = false;
          $originString .= $n . "=" . $v;
    }
    return $originString;
}

function getSignature(string $originString) {
    global $secretKey;
    $signature = hash_hmac('sha256', $originString, $secretKey, true);
    $signature = bin2hex($signature);
    return $signature;
}

function getRequestUrl(string $path, array $params) {
    global $url;
    $requestUrl = $url.$path."?";
    $first = true;
    foreach($params as $n => $v) {
          if (!$first) {
              $requestUrl .= "&";
          }
          $first = false;
          $requestUrl .= $n . "=" . $v;
    }
    return $requestUrl;
}

function httpPost($url)
{
    global $apiKey;
    $curl = curl_init($url);
    curl_setopt($curl, CURLOPT_POST, true);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_USERAGENT, "curl/7.80.0");
    curl_setopt($curl, CURLOPT_HTTPHEADER, array(
        "X-BX-APIKEY:".$apiKey,
    ));
    $response = curl_exec($curl);
    curl_close($curl);
    return $response;
}

function placeOrder(string $symbol, string $side, string $tradeType, string $quantity,
    string $quoteOrderQty, string $price) {

    // interface info
    $path = "/openApi/spot/v1/trade/order";

    // interface params
    $params = array();
    $params['symbol'] = $symbol;
    $params['side'] = $side;
    $params['type'] = $tradeType;
    $params['quantity'] = $quantity;
    $params['quoteOrderQty'] = $quoteOrderQty;
    $params['price'] = $price;
    $date = new DateTime();
    $params['timestamp'] = $date->getTimestamp()*1000;

    // generate signature
    $originString = getOriginString($params);
    $signature = getSignature($originString);
    $params["signature"] = $signature;

    // send http request
    $requestUrl = getRequestUrl($path, $params);
    $result = httpPost($requestUrl);
    echo "\t";
    echo $result;
    echo "\n";
}

function cancelOrder(string $symbol, string $orderId) {
    // interface info
    $path = "/openApi/spot/v1/trade/cancel";

    // interface params
    $params = array();
    $params['symbol'] = $symbol;
    $params['orderId'] = $orderId;
    $date = new DateTime();
    $params['timestamp'] = $date->getTimestamp()*1000;

    // generate signature
    $originString = getOriginString($params);
    $signature = getSignature($originString);
    $params["signature"] = $signature;

    // send http request
    $requestUrl = getRequestUrl($path, $params);
    $result = httpPost($requestUrl);
    echo "\t";
    echo $result;
    echo "\n";
}

function queryOrder(string $symbol, string $orderId) {
    // interface info
    $path = "/openApi/spot/v1/trade/query";

    // interface params
    $params = array();
    $params['symbol'] = $symbol;
    $params['orderId'] = $orderId;
    $date = new DateTime();
    $params['timestamp'] = $date->getTimestamp()*1000;

    // generate signature
    $originString = getOriginString($params);
    $signature = getSignature($originString);
    $params["signature"] = $signature;

    // send http request
    $requestUrl = getRequestUrl($path, $params);
    $result = httpPost($requestUrl);
    echo "\t";
    echo $result;
    echo "\n";
}

function queryOpenOrders(string $symbol) {
    // interface info
    $path = "/openApi/spot/v1/trade/openOrders";

    // interface params
    $params = array();
    $params['symbol'] = $symbol;
    $date = new DateTime();
    $params['timestamp'] = $date->getTimestamp()*1000;

    // generate signature
    $originString = getOriginString($params);
    $signature = getSignature($originString);
    $params["signature"] = $signature;

    // send http request
    $requestUrl = getRequestUrl($path, $params);
    $result = httpPost($requestUrl);
    echo "\t";
    echo $result;
    echo "\n";
}

function queryHistoryOrders(string $symbol, string $orderId, string $startTime,
    string $endTime, string $limit) {
    // interface info
    $path = "/openApi/spot/v1/trade/historyOrders";

    // interface params
    $params = array();
    $params['symbol'] = $symbol;
    $params['orderId'] = $orderId;
    $params['startTime'] = $startTime;
    $params['endTime'] = $endTime;
    $params['limit'] = $limit;
    $date = new DateTime();
    $params['timestamp'] = $date->getTimestamp()*1000;

    // generate signature
    $originString = getOriginString($params);
    $signature = getSignature($originString);
    $params["signature"] = $signature;

    // send http request
    $requestUrl = getRequestUrl($path, $params);
    $result = httpPost($requestUrl);
    echo "\t";
    echo $result;
    echo "\n";
}

// echo "placeOrder:\n";
// placeOrder("XRP-USDT", "BUY", "LIMIT", "20", "0", "0.5");

// echo "cancelOrder:\n";
// cancelOrder("XRP-USDT", "1513809777442947072");

// echo "queryOrder:\n";
// queryOrder("XRP-USDT", "1513809777442947072");

// echo "queryOpenOrders:\n";
// queryOpenOrders("XRP-USDT");

// echo "queryHistoryOrders:\n";
// queryHistoryOrders("XRP-USDT", "0", "0", "0", "2");

?>